File: app/src/main/java/at/shockbytes/corey/workout/model/TimeExercise.java
package at.shockbytes.corey.workout.model;

import android.content.Context;
import android.os.Parcel;
import android.support.annotation.NonNull;

import at.shockbytes.corey.R;

/**
 * @author Martin Macheiner
 *         Date: 20.03.2015.
 */
public class TimeExercise extends Exercise {

    // Both in seconds
    private int workDuration;
    private int restDuration;

    public TimeExercise() {
        this("", 0, 0);
    }

    public TimeExercise(String name) {
        this(name, 0);
    }

    public TimeExercise(String name, int reps) {
        super(name, reps);
    }

    public TimeExercise(String name, int reps, int workduration) {
        this(name, reps, workduration, workduration);
    }

    public TimeExercise(String name, int reps, int workduration, int restduration) {
        super(name, reps);
        workDuration = workduration;
        restDuration = restduration;
    }

    protected TimeExercise(Parcel in) {
        super(in);
        workDuration = in.readInt();
        restDuration = in.readInt();
    }

    public static final Creator<TimeExercise> CREATOR = new Creator<TimeExercise>() {
        @Override
        public TimeExercise createFromParcel(Parcel in) {
            return new TimeExercise(in);
        }

        @Override
        public TimeExercise[] newArray(int size) {
            return new TimeExercise[size];
        }
    };

    @Override
    public String getDisplayName(@NonNull Context context) {

        int minutes = getWorkoutDurationInSeconds() / 60;
        if (minutes != 0) {
            return context.getString(R.string.format_time_exercise, minutes, getName()).replaceAll("_", " ");
        } else {
            return getName().replaceAll("_", " ");
        }
    }

    public int getWorkoutDurationInSeconds() {
        return (repetitions * (workDuration + restDuration));
    }

    public int getWorkDuration() {
        return workDuration;
    }

    public void setWorkDuration(int workDuration) {
        this.workDuration = workDuration;
    }

    public int getRestDuration() {
        return restDuration;
    }

    public void setRestDuration(int restDuration) {
        this.restDuration = restDuration;
    }

    @Override
    public String toString() {
        return super.toString() + "\nWorkduration:\t" + workDuration + "\nRestduration:\t" + restDuration;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        super.writeToParcel(parcel, i);
        parcel.writeInt(workDuration);
        parcel.writeInt(restDuration);
    }

}

File: core/src/main/java/at/shockbytes/corey/common/core/workout/model/TimeExercise.kt
package at.shockbytes.corey.common.core.workout.model

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import at.shockbytes.corey.common.core.R
import com.google.firebase.database.Exclude

/**
 * Author:  Martin Macheiner
 * Date:    20.03.2015
 */
class TimeExercise(
    name: String,
    reps: Int = 0,
    eq: Equipment = Equipment.BODYWEIGHT,
    var workDuration: Int = 0,
    var restDuration: Int = workDuration
) : Exercise(name, reps, eq), Parcelable {

    val workoutDurationInSeconds: Int
        @Exclude
        get() = repetitions * (workDuration + restDuration)

    constructor(parcel: Parcel) : this(
            parcel.readString() ?: "",
            parcel.readInt(),
            parcel.readSerializable() as Equipment,
            parcel.readInt(),
            parcel.readInt())

    override fun getDisplayName(context: Context): String {

        val minutes = workoutDurationInSeconds / 60
        return if (minutes != 0) {
            context.getString(R.string.format_time_exercise, minutes, name).replace("_".toRegex(), " ")
        } else {
            name.replace("_".toRegex(), " ")
        }
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        super.writeToParcel(parcel, flags)
        parcel.writeInt(workDuration)
        parcel.writeInt(restDuration)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<TimeExercise> {
        override fun createFromParcel(parcel: Parcel): TimeExercise {
            return TimeExercise(parcel)
        }

        override fun newArray(size: Int): Array<TimeExercise?> {
            return ArrayList<TimeExercise>(size).toTypedArray()
        }
    }
}
